File: app/src/main/java/com/majeur/psclient/model/PokemonId.java
package com.majeur.psclient.model;

public class PokemonId {

    public static PokemonId fromRawId(Player player, String rawId) {
        return new PokemonId(player, rawId);
    }

    public static PokemonId fromPosition(Player player, int position) {
        return new PokemonId(player, position);
    }

    public static PokemonId mockNameOnly(String name) {
        return new PokemonId(name);
    }

    public final Player player;
    public final int position;
    public final String name;
    public final boolean foe;
    public final boolean isInBattle;

    PokemonId(Player player, String rawId) {
        this.player = player;
        position = rawId.charAt(2) - 'a';
        name = rawId.substring(rawId.indexOf(":") + 2);
        foe = player == Player.FOE;
        isInBattle = position >= 0;
    }

    PokemonId(String name) {
        this.player = null;
        this.name = name;
        position = -1;
        foe = false;
        isInBattle = false;
    }

    PokemonId(Player player, int position) {
        this.player = player;
        this.name = null;
        this.position = position;
        foe = player == Player.FOE;
        isInBattle = false;
    }

    @Override
    public String toString() {
        return "PokemonId{" +
                "player=" + player +
                ", position=" + position +
                ", name='" + name + '\'' +
                ", foe=" + foe +
                ", isInBattle=" + isInBattle +
                '}';
    }
}

File: app/src/main/java/com/majeur/psclient/model/battle/PokemonId.kt
package com.majeur.psclient.model.battle


class PokemonId {

    val player: Player
    val position: Int
    val name: String

    val foe get() = player == Player.FOE
    val trainer get() = player == Player.TRAINER
    val isInBattle get() = position >= 0


    // |switch|p1a: Quagsire|Quagsire, L83, M|100/100
    constructor(player: Player, rawId: String) {
        this.player = player
        position = rawId[2] - 'a' // char to alphabet index, if pkmn not in battle ':' - 'a' gives negative position
        name = rawId.substringAfter(": ")
    }

    constructor(player: Player, position: Int) {
        this.player = player
        this.position = position
        name = ""
    }

    override fun equals(other: Any?) = other is PokemonId && foe == other.foe && position == other.position
    override fun hashCode() = 31 * position + name.hashCode()
}
